### ISSUE TITLE
DiT blok seçenekleri: uzun skip, value residual, FFN depthwise conv, attention kapısı, SwiGLU, final adaLN

### ISSUE THREAD

**Issue by kadirnar** [kadirnar/dacvae-tts#9]
## Özet
Omurganın ana yapısı (self-attention + LARoPE'li cross-attention, paylaşımlı düşük-rank adaLN, QK-norm, patch 1)
kanıtlarla destekleniyor ve korunacak. Parametre maliyeti <1,5M olan, küçük modelde WER/kaliteyi artırma potansiyeli
taşıyan seçenekler konfigürasyonla açılıp kapanabilir şekilde eklenecek (varsayılan kapalı).

## Kanıt
| Seçenek | Kanıt | Maliyet |
|---|---|---|
| Giriş→çıkış uzun skip (concat + LN + Linear) | DiTTo (cross-attention DiT): WER 3,30 → 2,93, SIM-r 0,573 → 0,588; EzAudio: yakınsamayı hızlandırıyor. Ters: F5 (in-context) 4,17 → 5,17 | +0,5M |
| Value residual (v_l ← λ1·v_l + λ2·v_1) | 140M DiT'te (SR-DiT) FID 4,02 → 3,64; ResFormer 82M LM | ~0 |
| FFN içinde depthwise conv (k=5) | ZipVoice conv modülleri kaldırılınca WER 1,69 → 9,79; U-DiT/SANA kazanç; FastSpeech CMOS −0,11 (conv'suz). Ters: F5 Conv2Audio | +0,1M, ~%2 |
| Head-wise sigmoid attention kapısı | Qwen gated attention (LLM): kararlılık, yüksek LR toleransı; Echo/Irodori/Darya kullanıyor; TTS ablation'ı yok | +0,1M |
| SwiGLU/GEGLU (eşit parametre) | T5/LightningDiT pozitif, 140M SR-DiT nötr | 0 |
| Final adaLN + koşula havuzlanmış hedef metni | DiT/F5 standart; DiTTo havuzlanmış metin 3,00 → 2,93 | +0,3–0,8M |

Önerilmeyenler (eklenmeyecek): joint/MM-DiT attention, patch/downsampling, katman paylaşımı, differential attention.

## Kapsam (kod)
- `ModelConfig` alanları (hepsi varsayılan kapalı): `long_skip`, `value_residual`, `ffn_conv_kernel`, `attn_gate`
  (`none|head`), `ffn_activation` (`gelu|swiglu`), `final_adaln`, `cond_text_pool`.
- Yeni parametreler sıfır-init veya özdeşlik başlatmalı ki açık seçeneklerle model başlangıçta baz modelle aynı fonksiyonu
  hesaplasın (uygun olduğu yerde).
- `optim.py`: yeni matrislerin Muon/AdamW ayrımı ve kaynaşık matrislerin parça sayısı.
- Eski checkpoint'ler seçenekler kapalıyken `load_state_dict(strict=True)` ile yüklenir.
- Örnek konfigler: `configs/experiments/tr_w512_<seçenek>.yaml` (her biri tek değişken).

## Kabul ölçütleri
- [ ] Her seçenek için şekil/maske/ileri-geri testleri; parametre sayısı raporu.
- [ ] Seçenekler kapalıyken çıktı eskiyle birebir; eski state_dict yüklenir.
- [ ] Sıfır-init seçeneklerde başlangıç çıktısı baz ile aynı.

## Deney (GPU)
w512 baz 2 tohum + her seçenek tek başına, 20k güncelleme; Freya CER/WER (GA), SIM-o, DNSMOS/UTMOS, adım süresi.
CER kazancı tohum yayılımından büyük olanlar birleştirilir.

---
**Branch:** `arch/dit-blok-secenekleri` · **Araştırma raporu:** [arastirma-raporu-2026-09-24.md](https://github.com/kadirnar/dacvae-tts/blob/docs/temizlik-arastirma-raporu/docs/arastirma-raporu-2026-09-24.md)

**Comment by kadirnar** [comment]
## Uygulama: `arch/dit-blok-secenekleri`

Branch: [`arch/dit-blok-secenekleri`](https://github.com/kadirnar/dacvae-tts/tree/arch/dit-blok-secenekleri), `main` (8f01f08) üzerine 7 commit, seçenek başına bir commit. PR açılmadı.

### Seçenekler (hepsi varsayılan kapalı)

Parametre sayıları `dacvae-tts inspect` çıktısı. Baz `nano_tr_w512` **66.522.244** parametre.

| Seçenek | Config anahtarı (varsayılan → A/B değeri) | w512 maliyeti | Toplam | Başlangıç | Optimizer |
|---|---|---|---|---|---|
| Giriş→çıkış uzun skip | `long_skip: false` → `true` | +526.848 | 67.049.092 | sıfır-init: baz ile aynı | Linear(2D,D) Muon (1 parça), LN AdamW |
| Value residual | `value_residual: false` → `true` | +24 | 66.522.268 | λ1=1, λ2=0: baz ile aynı | AdamW (skaler) |
| FFN depthwise conv | `ffn_conv_kernel: 0` → `5` | +110.592 | 66.632.836 | sıfır-init: baz ile aynı | AdamW (3-B filtre) |
| Head-wise attention kapısı | `attn_gate: none` → `head` | +98.496 | 66.620.740 | 2·sigmoid(0)=1: baz ile aynı | AdamW (`ADAMW_SUFFIXES`) |
| SwiGLU (eşit parametre) | `ffn_activation: gelu` → `swiglu` | +6.144 | 66.528.388 | yeni FFN (sıfır-init değil) | Muon, gate/value 2 parça (`FUSED_ROWS`) |
| Final adaLN | `final_adaln: false` → `true` | +99.392 (rank 64) | 66.621.636 | sıfır-init: baz ile aynı | Muon; shift/scale 2 parça |
| Koşula havuzlanmış metin | `cond_text_pool: false` → `true` | +262.144 | 66.784.388 | sıfır-init: baz ile aynı | Muon (1 parça) |

Hepsi birden açıkken: 67.589.020 (+1.066.776).

### Uygulama notları
- **long_skip:** son blok çıkışından sonra `h = h_L + Linear0(LN([h_0, h_L]))`. `h_0` giriş projeksiyonunun çıktısı. Toplamı çıkış başının LayerNorm'u normalize ediyor; Hunyuan-DiT de füzyondan sonra norm kullanıyor.
- **value_residual:** self-attention'da SDPA öncesi head başına değerler `v_l ← λ1·v_l + λ2·v_1`. `v_1` blok çağrısına açık argüman olarak veriliyor ve geri dönüyor, bu yüzden activation checkpointing ile uyumlu; checkpoint açık ve kapalı gradyanlar aynı. İlk blok kendi değerleriyle karışıyor, böylece DDP'de kullanılmayan parametre kalmıyor.
- **ffn_conv_kernel:** FFN gizli katmanında aktivasyondan sonra residual depthwise Conv1d var. Conv'dan önce padding kareleri sıfırlanıyor: bias'tan gelen aktivasyonlar komşu geçerli karelere sızmasın diye. Test ediliyor, patch_size 2 dahil. Çift ve negatif kernel reddediliyor.
- **attn_gate:** üretici bloklarında self- ve cross-attention'a uygulanıyor, metin kodlayıcı değişmedi. Kapı sorgu tarafındaki girişten hesaplanıyor: `y_h ← 2·sigmoid(w_h·x + b_h)·y_h`. Sıfır-init ile kapı tam 1'den başlıyor ve (0, 2) aralığında öğreniliyor.
- **swiglu:** gizli boyut GELU'nun 2/3'ü, 64'ün katına yuvarlanıyor (512×3 için 1024). Tek bir kaynaşık projeksiyon var; Muon gate ve value yarılarını ayrı ortogonalize ediyor. `ffn_conv_kernel` ile birlikte kullanılabiliyor.
- **final_adaln:** çıkış LayerNorm'unun shift/scale değerleri koşuldan (zaman + ses) geliyor. `adaln_rank > 0` iken bloklardaki gibi rank-r, aksi halde tam D→2D (w512'de +525.312 olurdu).
- **cond_text_pool:** `cond += Linear0(hedef byte kodlamalarının maskeli ortalaması)`. Duration başıyla aynı byte'lar kullanılıyor; joined düzende bu tüm transkript demek. Bias yok ve ortalama koşul dropout'undan sonra alınıyor. Bu yüzden CFG null dalında terim tam sıfır; eğitim, `sample` içindeki toplu null dal ve text-only dal tutarlı kalıyor. Önbellek sözleşmesi `(text, text_valid, voice)` değişmedi.
- Sıfır-init modüller RNG tüketmeden kuruluyor (`nn.utils.skip_init`), FlowTTS düzeyindekiler de tüm baz modüllerden sonra kuruluyor. Sonuç: **aynı tohumla her A/B kolu bazla birebir aynı ağırlıklardan başlıyor.** Model kurulduktan sonra tohum yeniden verildiği için gürültü, zaman ve dropout çekilişleri de aynı. Kollar arasındaki tek fark seçeneğin kendisi.

### Seçenekler kapalıyken
- Modüller, `state_dict` anahtarları ve hesap değişmedi. 8f01f08'e karşı yakalanan karşılaştırmada (nano, nano+rank, plain, 4 bloklu w512) başlangıç ağırlıkları, eval çıktısı, eğitim kaybı ve tüm gradyanlar bit düzeyinde aynı (`torch.equal`).
- Kalıcı regresyon testi: öncekinin `state_dict` düzen hash'i ve RNG'siz bir ileri geçiş parmak izi. Eski checkpoint'ler `strict=True` ile yükleniyor, seçenekler açıkça `false` verildiğinde de.

### Testler
- Yeni `tests/test_block_options.py`, 33 test:
  - her seçenek için ileri/geri geçiş, checkpoint açık ve kapalı gradyan eşitliği, padding değişmezliği;
  - sıfır-init seçeneklerde başlangıç çıktısının bazla `torch.equal` olması ve aynı tohumda baz ağırlıkların aynı kalması;
  - Muon/AdamW bölümlemesi ve parça sayıları, parametre sayıları, config doğrulaması;
  - seçeneğe özgü anlamsal testler: λ=(0,1) iken blok 2'nin kendi value projeksiyonunun etkisiz olması, kapı ile head kapatma, scale=−1 ile final adaLN, havuzun yalnız hedef byte'ları alması;
  - tüm seçenekler açıkken CFG null dalında transkriptin hiçbir yoldan sızmaması;
  - örnek konfiglerin w512'den yalnız tek anahtarla ayrılması.
- Tam paket (bu makinede ortam kaynaklı başarısız olan iki DDP subprocess testi `--deselect` ile hariç): **180 passed, 3 skipped**. `ruff check` temiz. bf16 autocast ve `torch.compile` (CPU) ile tüm seçenekler açık da denendi; compile çıktısı eager ile aynı.

### A/B planı (GPU)
Örnek konfigler `nano_tr_w512.yaml`'ın birebir kopyası, tek fark bir model anahtarı. Testler bunu doğruluyor.

| Kol | Config |
|---|---|
| Baz, tohum 42 | `configs/nano_tr_w512.yaml` |
| Baz, tohum 43 (tohum yayılımı) | `configs/experiments/tr_w512_seed43.yaml` |
| Uzun skip | `configs/experiments/tr_w512_long_skip.yaml` |
| Value residual | `configs/experiments/tr_w512_value_residual.yaml` |
| FFN conv (k=5) | `configs/experiments/tr_w512_ffn_conv.yaml` |
| Attention kapısı | `configs/experiments/tr_w512_attn_gate.yaml` |
| SwiGLU | `configs/experiments/tr_w512_swiglu.yaml` |
| Final adaLN | `configs/experiments/tr_w512_final_adaln.yaml` |
| Havuzlanmış metin | `configs/experiments/tr_w512_cond_text_pool.yaml` |

- Her kol aynı 60k LR programıyla 20k güncelleme koşacak: `train --stop-after 20000`, programı değiştirmeden durduruyor. Aynı frame budget kullanılacak, örneğin:
  `TRAIN_GPUS=1 bash scripts/train_2gpu.sh configs/experiments/tr_w512_long_skip.yaml <cache> runs/tr-w512-long-skip --frame-budget 6000 --stop-after 20000`
- Ölçümler 20k'da: Freya CER/WER (GA), SIM-o, DNSMOS/UTMOS, adım süresi.
- Karar kuralı: CER kazancı baz 42/43 yayılımından büyük olan seçenekler birleştirilecek, sonra birleşik kol tek koşu olarak doğrulanacak.

### Riskler ve notlar
- FFN conv'un "~%2 hesap" maliyeti araştırma tahmini; gerçek adım süresi GPU'da ölçülmeli.
- Attention kapısında 2·sigmoid ile başlangıç 1. Qwen'de kapılar seyrek değerlere iniyor; burada o ön bilgi, başlangıçta baza tam eşitlik için bırakıldı.
- λ'lar ve kapılar kod tabanının geleneğine uygun olarak AdamW'de weight decay alıyor.
- Uzun skip in-context F5'te kötüleşmişti (4,17 → 5,17). Bizim düzen DiTTo gibi cross-attention; yine de A/B şart.
